core/embed/io/nrf/stm32u5/nrf_update.c

```diff
@@ -87,45 +80,118 @@ static int read_image_sha256(const uint8_t *binary_ptr, size_t binary_size,
+  return ret;
+}
+
+/**
+ * Read the image version from the image header.
+ *
+ * @param image_ptr  pointer to the binary image
+ * @param out_version   Pointer to nrf_app_version_t to receive the version
+ * @return "true" on success, "false" on failure
+ */
+static bool image_version_read(const uint8_t *image_ptr,
+                               nrf_app_version_t *out_version) {
+  struct image_header *hdr = (struct image_header *)image_ptr;
+
+  if (image_ptr == NULL || out_version == NULL) {
+    return false;
+  }
+
+  memcpy(out_version, &hdr->ih_ver, sizeof(nrf_app_version_t));
+
+  return true;
+}
+
+/**
+ * Read the image version from the nRF MCUboot via SMP serial recovery.
+ *
+ * @param out_version   Pointer to nrf_app_version_t to receive the version
+ * @return "true" on success, "false" on failure
+ */
+static bool nrf_smp_version_get(nrf_app_version_t *out_version) {
+  bool ret = false;
+
+  nrf_reboot_to_bootloader();
+  nrf_set_dfu_mode(true);
+
+  if (smp_image_version_get(out_version)) {
+    // Success - version string provided via SMP has been decoded and stored
+    // within "out_version" variable
+    ret = true;
+  }
+
+  nrf_reboot();
+  nrf_set_dfu_mode(false);
+
+  return ret;
+}
+
+/**
+ * Comparison of two image versions.
+ *
+ * @param v1 Pointer to first nrf_app_version_t
+ * @param v2 Pointer to second nrf_app_version_t
+ * @return 0 when equal, 1 when v1 is greater, -1 when v2 is greater
+ */
+static int version_cmp(const nrf_app_version_t *v1,
+                       const nrf_app_version_t *v2) {
+  if (v1->major != v2->major) {
+    return (v1->major < v2->major) ? -1 : 1;
+  }
+  if (v1->minor != v2->minor) {
+    return (v1->minor < v2->minor) ? -1 : 1;
+  }
+  if (v1->revision != v2->revision) {
+    return (v1->revision < v2->revision) ? -1 : 1;
+  }
+  if (v1->build_num != v2->build_num) {
+    return (v1->build_num < v2->build_num) ? -1 : 1;
+  }
+  return 0;
 }
 
 bool nrf_update_required(const uint8_t *image_ptr, size_t image_len) {
-  nrf_info_t info = {0};
+  for (int i = 0; i < 3; i++) {
+    nrf_info_t info;
+    uint8_t expected_hash[SHA256_DIGEST_LENGTH];
 
-  uint16_t try_cntr = 0;
-  while (!nrf_get_info(&info)) {
-    nrf_reboot();
-    systick_delay_ms(500);
-    try_cntr++;
-    if (try_cntr > 3) {
-      // Assuming corrupted image, but we could also check comm with MCUboot
-      return true;
+    if (nrf_get_info(&info) == true &&
+        read_image_sha256(image_ptr, image_len, expected_hash) == true) {
+      return memcmp(info.hash, expected_hash, SHA256_DIGEST_LENGTH) != 0;
     }
-  }
 
-  uint8_t expected_hash[SHA256_DIGEST_LENGTH] = {0};
+    // Can't communicate with the App via SPI, trying SMP serial recovery over
+    // UART to nRF MCUboot
+    nrf_app_version_t smp_version, image_version;
+
+    if (nrf_smp_version_get(&smp_version) == true &&
+        image_version_read(image_ptr, &image_version) == true) {
+      return version_cmp(&image_version, &smp_version) > 0;
```

Comment:
for maintaining compatibility between firmware & nRF, i think its better to also allow downgrades by comparing with `!=`. Hash comparison above also behaves this way